File: android/app/src/main/java/com/foodsense/android/services/StreakManager.kt
package com.foodsense.android.services

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

class StreakManager(
    private val context: Context,
    private val nutritionManager: NutritionManager,
) {
    private val prefs = context.getSharedPreferences("streak_prefs", Context.MODE_PRIVATE)

    var currentStreak by mutableStateOf(0)
        private set

    var longestStreak by mutableStateOf(prefs.getInt("longestStreak", 0))
        private set

    init {
        updateStreak()
    }

    /** Recalculate the current streak from NutritionManager logs. */
    fun updateStreak() {
        currentStreak = calculateStreak(nutritionManager.logs)

        if (currentStreak > longestStreak) {
            longestStreak = currentStreak
            prefs.edit().putInt("longestStreak", longestStreak).apply()
        }
    }

    companion object {
        /** Walk backwards from today counting consecutive days with at least one log. */
        fun calculateStreak(logs: List<com.foodsense.android.data.FoodLog>): Int {
            val loggedDays: Set<LocalDate> = logs.map {
                Instant.ofEpochMilli(it.timeEpochMillis)
                    .atZone(ZoneId.systemDefault())
                    .toLocalDate()
            }.toSet()

            var streak = 0
            var checkDate = LocalDate.now()

            while (loggedDays.contains(checkDate)) {
                streak++
                checkDate = checkDate.minusDays(1)
            }

            return streak
        }
    }
}

File: FoodDetectionApp/StreakManager.swift
import Foundation
import Combine

class StreakManager: ObservableObject {
    static let shared = StreakManager()

    @Published var currentStreak: Int = 0
    @Published var longestStreak: Int = 0

    private let longestStreakKey = "longestStreak"

    init() {
        longestStreak = UserDefaults.standard.integer(forKey: longestStreakKey)
        updateStreak()
    }

    /// Recalculate the current streak from NutritionManager logs.
    func updateStreak() {
        let logs = NutritionManager.shared.logs
        currentStreak = Self.calculateStreak(from: logs)

        if currentStreak > longestStreak {
            longestStreak = currentStreak
            UserDefaults.standard.set(longestStreak, forKey: longestStreakKey)
        }

        BadgeManager.shared.checkBadges()
    }

    // MARK: - Streak Calculation

    /// Walk backwards from today and count consecutive days that have at least one log.
    static func calculateStreak(from logs: [NutritionManager.FoodLog]) -> Int {
        let calendar = Calendar.current

        // Build a set of unique logged days (normalised to start-of-day).
        let loggedDays: Set<Date> = Set(logs.map { calendar.startOfDay(for: $0.time) })

        var streak = 0
        var checkDate = calendar.startOfDay(for: Date())

        while loggedDays.contains(checkDate) {
            streak += 1
            guard let previous = calendar.date(byAdding: .day, value: -1, to: checkDate) else { break }
            checkDate = previous
        }

        return streak
    }

}
